apps/src/music/utils/UniqueSounds.js

```diff
@@ -0,0 +1,75 @@
+export default class UniqueSounds {
+  constructor() {
+    this.previousUniqueSounds = null;
+  }
+
+  // Given some song data events, this function returns an array of the
+  // sound IDs in the order that we want to render them in the timeline UI.
+  // If available, it uses its own previous return value to ensure that
+  // sounds remain in the same row as last time.
+  //
+  // It does remove empty rows, so if a sound is removed, then everything
+  // can shift up.
+  //
+  // If a sound is replaced in subsequent calls, which will happen when a
+  // block has its sound changed in the dropdown, then the stickiness of
+  // all other sounds means that the new sound will replace the old one in
+  // the same row.
+
+  getUniqueSounds(songDataEvents) {
+    // Each unique sound gets its own row (and therefore color).
+    // If a sound was showing previously, then we'll attempt to keep
+    // it in the same row, though we won't have empty rows.
+
+    // First, generate a list of all unique sounds.
+    const uniqueSounds = [];
+    for (const songEvent of songDataEvents) {
+      const id = songEvent.id;
+      if (uniqueSounds.indexOf(id) === -1) {
+        uniqueSounds.push(id);
+      }
+    }
+
+    // This is the actual output.  It will be the same length as uniqueSounds.
+    // Fill it with undefined entries.
+    let outputSounds = new Array(uniqueSounds.length).fill(undefined);
+
+    // If we have a previous output from this function, then we'll attempt
+    // to keep those entries in the same rows as they were before.
+    if (this.previousUniqueSounds) {
+      // For each of those previous entries...
+      for (let i = 0; i < this.previousUniqueSounds.length; i++) {
+        // ...if it's still in use...
+        if (uniqueSounds.indexOf(this.previousUniqueSounds[i]) !== -1) {
+          // ... put it back in that row.
+          outputSounds[i] = this.previousUniqueSounds[i];
```

Comment:
Could you now remove this sound from `uniqueSounds` so that on the next pass below you don't have to check if the sound has already been added?

Reply:
That's a neat idea.  One thing I have been attempting to do is to keep the data somewhat immutable, so I wanted to leave that prior array intact while we construct the new one, though it does mean we have to do some extra checking.